test/e2e-middleware-error.test.ts

```diff
@@ -0,0 +1,95 @@
+import { NextApiRequest, NextApiResponse } from 'next';
+import request from 'supertest';
+import { BadRequestException, createHandler, Get, UseBefore } from '../lib';
+import { NextFunction } from '../lib/decorators';
+import { setupServer } from './setupServer';
+
+const messages: string[] = [];
+
+@UseBefore(
+  (req: NextApiRequest, res: NextApiResponse, next: NextFunction) => {
+    if (req.url?.includes('/will-throw')) {
+      throw new Error('An error occurred.');
+    } else if (req.url?.includes('/bad-request')) {
+      throw new BadRequestException();
+    } else if (req.url?.includes('/return-error')) {
+      return next(new BadRequestException());
+    }
+
+    res.status(400).end();
+    next();
+  },
+  (_: any, __: any, next: NextFunction) => {
+    messages.push('A message.');
+    next();
+  }
+)
+class TestHandler {
+  @Get()
+  public index(): string {
+    return 'Hello!';
+  }
+
+  @Get('/will-throw')
+  public willThrow(): string {
+    return 'An error!';
+  }
+
+  @Get('/bad-request')
+  public badRequest(): string {
+    return 'Bad Request';
+  }
+
+  @Get('/return-error')
+  public returnError(): null {
+    return null;
+  }
+}
+
+describe('E2E - Middleware - Errors', () => {
+  let server: ReturnType<typeof setupServer>;
+  beforeAll(() => (server = setupServer(createHandler(TestHandler))));
+  afterAll(() => {
+    if ('close' in server && typeof server.close === 'function') {
+      server.close();
+    }
+  });
+
+  it('Should not run the other middlewares when a response is written in the previous middleware.', async () => {
```

Comment:
Code wise the PR looks perfect, amazing new functionality with a lot of use cases 💪 

This is the only part I'm doubting functional wise (not executing the next middleware(s) if a response has been sent). I think throwing or returning without calling `next` should be the only way to break the middleware execution chain. It will give a bit more responsibility to the developer to (not) handle the response properly, but it also enables more flexibility functional wise because it puts the developer in full control of the flow to either execute or not execute the next middleware function.